crates/deltalake-aws/src/storage.rs

```diff
@@ -0,0 +1,597 @@
+//! AWS S3 storage backend.
+
+use bytes::Bytes;
+use deltalake_core::storage::object_store::{
+    aws::AmazonS3ConfigKey, parse_url_opts, GetOptions, GetResult, ListResult, MultipartId,
+    ObjectMeta, ObjectStore, Result as ObjectStoreResult,
+};
+use deltalake_core::storage::{str_is_truthy, ObjectStoreFactory, ObjectStoreRef, StorageOptions};
+use deltalake_core::{DeltaResult, ObjectStoreError, Path};
+use futures::stream::BoxStream;
+use rusoto_core::Region;
+use std::collections::HashMap;
+use std::fmt::Debug;
+use std::ops::Range;
+use std::str::FromStr;
+use std::sync::Arc;
+use std::time::Duration;
+use tokio::io::AsyncWrite;
+use url::Url;
+
+const STORE_NAME: &str = "DeltaS3ObjectStore";
+
+#[derive(Clone, Default, Debug)]
+pub struct S3ObjectStoreFactory {}
+
+impl S3ObjectStoreFactory {
+    fn with_env_s3(&self, options: &StorageOptions) -> StorageOptions {
```

Comment:
The specifics of why this code does things these way are a little outside of my realm of understanding :laughing: 

I was _mostly_ moving code around, with some refactoring here and there. This code is largely doing the same thing it did when it lived in deltalake-core.  That said, this takes `StorageOptions` rather than `S3StorageOptions` because the result of this is intended to be passed into the `ObjectStoreFactory.parse_url_opts` which is a more generic trait and just operates on the `StorageOptions` struct